<div class="alert alert-block alert-success">
    
## Data Science Libraries Quiz

### Instructions


1. Write code which will produce the specified result  
2. Enter that result in the green response cell when prompted
3. Write code comment(s) to describe the code


#### Note:

1. Complete the work without the help of others
2. Enter one code 
comment on Question 0 for 1 point 
3. There are 9 questions, each worth 11 points
    - 10 points for code and answers
    - 1 point for at least one code comment on each question
4. You can receive partial credit for your code





<div class="alert alert-block alert-info">

### Question 0 

- Run the following two cells
- Enter one code comment

In [1]:
from pydataset import data
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = data('tips')#This is a comment for credit

---

<div class="alert alert-block alert-info">
  
## Question 1
        
a) What is the highest total_bill?
    
b) What time of day did this total_bill occur? 

In [12]:
df.head()

total_bill   tip     sex smoker  day    time  size
1       16.99  1.01  Female     No  Sun  Dinner     2
2       10.34  1.66    Male     No  Sun  Dinner     3
3       21.01  3.50    Male     No  Sun  Dinner     3
4       23.68  3.31    Male     No  Sun  Dinner     2
5       24.59  3.61  Female     No  Sun  Dinner     4

In [41]:
df.total_bill.nlargest(1)

171    50.81
Name: total_bill, dtype: float64

In [38]:
df[df.total_bill == df.total_bill.max()]

total_bill   tip   sex smoker  day    time  size
171       50.81  10.0  Male    Yes  Sat  Dinner     3

<div class="alert alert-block alert-success">

### Question 1 Response
    
a) The largest bill is $50.81
    
b) Dinner


---

<div class="alert alert-block alert-info">
  
## Question 2
    
a) What values are in the smoker column?
    
b) Are the majority of tables smoking or non-smoking?

In [43]:
df.info() # I used the .info() method to identify the types of values present in each column.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 244 entries, 1 to 244
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 23.4+ KB


In [49]:
df.value_counts('smoker')

smoker
No     151
Yes     93
dtype: int64

<div class="alert alert-block alert-success">

### Question 2 Response
    
a) Objects

b) There were a minority of smokers in this dataset.

---

<div class="alert alert-block alert-info">
  
## Question 3 
    
- What proportion of the bills are from the weekend (Saturday and/or Sunday)? 
    - Assume each table has one bill, regardless of table size
- Round your answer to two decimal points.

In [50]:
df.head()

total_bill   tip     sex smoker  day    time  size
1       16.99  1.01  Female     No  Sun  Dinner     2
2       10.34  1.66    Male     No  Sun  Dinner     3
3       21.01  3.50    Male     No  Sun  Dinner     3
4       23.68  3.31    Male     No  Sun  Dinner     2
5       24.59  3.61  Female     No  Sun  Dinner     4

<div class="alert alert-block alert-success">

### Question 3 Response
    
- 



---
<div class="alert alert-block alert-info">
    
## Question 4

    
-  Create a column named `percent_tip`.


- Calculate the percent_tip based on the formula below. 



> $\texttt{Percent tip} = \dfrac{\texttt{tip amount}}{\texttt{bill}}$

- What was the highest percent tip?

In [52]:
df['percent_tip'] = df['tip'] / df['total_bill']

In [54]:
df['percent_tip'].nlargest(1) * 100 #I used the above created column and then sorted it using .nlargest(1) to find the highest tip number. I made it a percentage by * 100.

173    71.034483
Name: percent_tip, dtype: float64

<div class="alert alert-block alert-success">

### Question 4 Response
    
- The highest percent tip was 71.03%
    

---

<div class="alert alert-block alert-info">
    
## Question 5

    
a) Which time of day has a higher average total_bill, dinner or lunch?
    
b) What is the difference in the average bill?


In [108]:
df.head()


total_bill   tip     sex smoker  day    time  table_size  percent_tip
1       16.99  1.01  Female     No  Sun  Dinner           2     0.059447
2       10.34  1.66    Male     No  Sun  Dinner           3     0.160542
3       21.01  3.50    Male     No  Sun  Dinner           3     0.166587
4       23.68  3.31    Male     No  Sun  Dinner           2     0.139780
5       24.59  3.61  Female     No  Sun  Dinner           4     0.146808

In [116]:
df.groupby('time').total_bill.mean().sort_values(ascending=False)

time
Dinner    20.797159
Lunch     17.168676
Name: total_bill, dtype: float64

In [90]:

df.time.sort_values(by=[['total.bill'].mean()]) # I sure hope that this got the right answer


AttributeError: 'list' object has no attribute 'mean'

<div class="alert alert-block alert-success">

### Question 5 Response
    
a) Dinner

b) 3.63

---
<div class="alert alert-block alert-info">
    
## Question 6
    

- Rename the column named `size` to `table_size`
- Verify this column name replaces the former column name 
- Use your new column name to determine:
    - What is the most common table size?


In [93]:
df = df.rename(columns = {'size': 'table_size'}) # I used .rename() to rename size to table_size
df

total_bill   tip     sex smoker   day    time  table_size  percent_tip
1         16.99  1.01  Female     No   Sun  Dinner           2     0.059447
2         10.34  1.66    Male     No   Sun  Dinner           3     0.160542
3         21.01  3.50    Male     No   Sun  Dinner           3     0.166587
4         23.68  3.31    Male     No   Sun  Dinner           2     0.139780
5         24.59  3.61  Female     No   Sun  Dinner           4     0.146808
..          ...   ...     ...    ...   ...     ...         ...          ...
240       29.03  5.92    Male     No   Sat  Dinner           3     0.203927
241       27.18  2.00  Female    Yes   Sat  Dinner           2     0.073584
242       22.67  2.00    Male    Yes   Sat  Dinner           2     0.088222
243       17.82  1.75    Male     No   Sat  Dinner           2     0.098204
244       18.78  3.00  Female     No  Thur  Dinner           2     0.159744

[244 rows x 8 columns]

In [117]:
df.table_size.sort_values(ascending=False).head() #I used .head() to find the top 5 most common table sizes to ID if I needed to expand my search for the top size.

144    6
157    6
126    6
142    6
186    5
Name: table_size, dtype: int64

<div class="alert alert-block alert-success">

### Question 6 Response
    
- 6
    

---
<div class="alert alert-block alert-info">
    
## Question 7

- The following code produces errors.
    
- Correct the code to determine:
    - What is the **maximum total_bill** for dinner time on Thursday?
    

#### Incorrect code
```python
df.total_bill( (df.day == 'Thursday') and (df.time == 'dinner') ) 
```

In [ ]:
df.sort_value()

<div class="alert alert-block alert-success">

### Question 7 Response

    
- 
    
    

---
<div class="alert alert-block alert-info">
    
## Question 8


- What is the min, mean, and max total_bill?
- Use a **histogram** to plot the distribution of the total_bill
    - add a title
    

In [ ]:
df['total_bill'].agg

<div class="alert alert-block alert-success">

### Question 8 Response

    
- 
    
    

---
<div class="alert alert-block alert-info">
    
## Question 9

- Use matplotlib to create a **scatter plot** for total_bill versus tips
    - Plot total_bill along the x-axis

- Customize the visualization in at least one way
- Add axis labels and a title